design/process-parameters.md

```diff
@@ -0,0 +1,230 @@
+steps:
+- script: echo Running one leg of tests...
+```
+ 
+### Select configs to run
+
+```yaml
+parameters:
+- name: configs
+  type: string
+  default: x86,x64
+  runTime: true
+
+jobs:
+- ${{ if contains(parameters.configs, 'x86' }}:
+  - job: x86
+    steps:
+    - script: echo Building x86...
+- ${{ if contains(parameters.configs, 'x64' }}:
+  - job: x64
+    steps:
+    - script: echo Building x64...
+- ${{ if contains(parameters.configs, 'arm' }}:
+  - job: arm
+    steps:
+    - script: echo Building arm...
+```
+
+*There's a possible future enhancement hiding here: a data type of "flags" which would be like an enum but allow multiple specification.
+This version is more verbose but also clear and straightforward.*
+
+### Select your pool
+
+```yaml
+parameters:
+- name: pool
+  type: pool
+  default: ubuntu-16.04
+  runTime: true
+
+job: Build
+pool:
+  vmImage: ${{ parameters.pool }}
+steps:
+- script: echo Building on the pool of your choice
+```
+
+### Connecting parameters with strong typing
+
+This is common for sophisticated customers such as .NET.
+
+```yaml
+# template.yml - comes from the centralized infrastructure team
+
+parameters:
+- name: customerSteps
+  type: steps
+
+steps:
+- script: echo Pre-step injected on all jobs
+- ${{ each customerStep in parameters.customerSteps }}:
+  - ${{ customerStep }}
+- script: echo Post-step injected on all jobs
+```
+
+When a client repo such as Roslyn goes to consume this template:
+
+```yaml
+# azure-pipelines.yml - in the Roslyn repo
+
+resources:
+  repositories:
+  - repository: arcade
+    name: dotnet/arcade
+
+steps:
+- template: template.yml@arcade
+  parameters:
+    customerSteps:
+    - task: Foo@1
+    - script: echo bar
+```
+
+At runtime, we know that `customerSteps` must be a list of steps.
+If it's not, we can give a clear error message.
+In the future, our editing experiences can also make this easy to reason about.
+
+## Back-compat with current syntax
+
+The existing `parameters` syntax expects a map of parameters with their defaults.
+Since the value of a parameter can be anything including a map, we wouldn't be able to tell whether the customer meant to use the old syntax or the new, richer syntax.
+That's why the new syntax is a sequence of defined objects, allowing us to preserve back-compat.
+
+Example of current syntax:
+```yaml
+parameters:
+  image: 'ubuntu-16.04'   # scalar parameter
+  preSteps:               # sequence parameter
+  - script: echo foo
+  vars:                   # map parameter
+    NODE_VERSION: '8.x'
+    PYTHON_VERSION: '3.7'
+```
+
+Since existing parameter syntax is only for templates, we assume none of the parameters are settable at run time.
+The above example is equivalent to:
+```yaml
+parameters:
+- name: 'image'     # scalar parameter
+  type: string
+  default: 'ubuntu-16.04'
+- name: 'preSteps'  # sequence parameter
+  type: object
+  default:
+  - script: echo foo
+- name: 'vars'      # map parameter
+  type: object
```

Comment:
Do we ever let a customer attach a schema here to define what the object should look like?

Reply:
Sounds like a good future enhancement